### PR TITLE
oneMKL: support axpy from level1

### PR DESCRIPTION
**Pull request by kballeda**
This PR contains axpy primitive from level 1
Closed following PR - [#237 ](https://github.com/JuliaGPU/oneAPI.jl/pull/237)

### CODE REVIEW

**Comment by maleadt** [comment]
Please *do not* open new PRs all the time; just make changes to your original PR's branch and push those.

**Comment by maleadt** [comment]
Also, please reduce the number of commits. By pushing "NITS" commits that change nothing but whitespace, CI time is needlessly wasted.

**Comment by kaliudayintel** [comment]
> Please do not open new PRs when there's an existing one; just make changes to your original PR's branch and push those.
> 
> Other than that, the CI failure here looks related, but happens during execution of `iamax`...

@maleadt  Thanks you. Sure, actually I happened to reopen a fresh PR as my local setup got corrupted after your merge of iamax during rebase/merge. I tried force pushing changes but that didnt work so raised a fresh PR to overcome it. Since this branch is created on top of my working version we have the history of CI runs here. I will ensure same PR maintained from next time.

**Comment by kaliudayintel** [comment]
> Also, please reduce the number of commits. By pushing "NITS" commits that change nothing but whitespace, CI time is needlessly wasted.

Sure will follow the same. This time I wanted to trigger CI run once again and found NITS to enable it.

**Comment by maleadt** [comment]
Thanks!

> my local setup got corrupted after your merge of iamax during rebase/merge. I tried force pushing changes but that didnt work so raised a fresh PR to overcome it

It requires a bit of git fu, but it's always possible to go to a new branch and then just push it to the old PR using `git push -u origin HEAD:your_previous_branch`.

**Comment by kaliudayintel** [comment]
@maleadt I find that in local testing these tests are passing and also couple of previous CI runs are successful.
image

**Comment by maleadt** [comment]
> I find that in local testing these tests are passing

Then more effort is required to reproduce locally. Try running the operations in a loop instead of re-starting the test process, that should make it possible to quickly run a whole bunch of iterations.

File: deps/src/onemkl.cpp [hunk 1]
```diff
@@ -81,6 +81,39 @@ extern "C" int onemklZgemm(syclQueue_t device_queue, onemklTranspose transA,
...
+                        reinterpret_cast<std::complex<float> *>(y), incy);
+    status.wait();
+}
+
+extern "C" void onemklZaxpy(syclQueue_t device_queue, int64_t n,
+                            double _Complex alpha, const double _Complex *x,
+                            std::int64_t incx, double _Complex *y, int64_t incy) {
+    auto status = oneapi::mkl::blas::column_major::axpy(device_queue->val, n,
+                        alpha, reinterpret_cast<const std::complex<double> *>(x), incx,
+                        reinterpret_cast<std::complex<double> *>(y), incy);
```
Comment (Sarbojit2019) [inline deps/src/onemkl.cpp:114]: wait() is missing here.

**Review by Sarbojit2019 — CHANGES_REQUESTED** [review]

**Comment by maleadt** [comment]
> Then more effort is required to reproduce locally. 

For something more productive: try the approach from https://github.com/JuliaGPU/oneAPI.jl/issues/243, where you include the test in a while loop. That allows you to execute much more iterations much quicker, because it doesn't have to recompile over and over.

**Comment by kaliudayintel** [comment]
> > Then more effort is required to reproduce locally.
> 
> For something more productive: try the approach from #243, where you include the test in a while loop. That allows you to execute much more iterations much quicker, because it doesn't have to recompile over and over.

@maleadt  Thanks for your suggestion. I have done this analysis and found that axpy fails locally in iterative runs. 
I will send a fix for the same soon.

**Comment by maleadt** [comment]
Why does this `wait()`? It shouldn't, because that means these operations would synchronize the GPU. Instead, we only synchronize when copying data back to the GPU.

`iamin` and `iamax` were different in that regard because it returned a scalar on the CPU, so there's no point where we can synchronize.

**Comment by pengtu** [comment]
> > Why does this `wait()`? It shouldn't, because that means these operations would synchronize the GPU. Instead, we only synchronize when copying data back to the GPU.
> > `iamin` and `iamax` were different in that regard because it returned a scalar on the CPU, so there's no point where we can synchronize.
> 
> The following test case fails spontaneously without the wait():
> 
> @test testf(axpy!, alpha[1], rand(T,m), rand(T,m))
> 
> Inside the "testf" function, it copies the GPU arrays back to the CPU. Your question indicates that a synchronization is already there. Would you point us to the synchronization code?
> 
> The oneMKL library call always return a SYCL event, do we need to hook it up into oneAPI.jl's dependency tracking module?
> 
> Another scenario is that the spontaneous failure is caused by running out of memory that you have recently fixed, unlikely but possible.

I can see the queue synchronization command at: 
https://github.com/JuliaGPU/oneAPI.jl/blob/master/src/array.jl#L322

Let's run ze_tracer to see the sequence of L0 calls when it failed.

**Comment by pengtu** [comment]
> > > Why does this `wait()`? It shouldn't, because that means these operations would synchronize the GPU. Instead, we only synchronize when copying data back to the GPU.
> > > `iamin` and `iamax` were different in that regard because it returned a scalar on the CPU, so there's no point where we can synchronize.
> > 
> > 
> > The following test case fails spontaneously without the wait():
> > @test testf(axpy!, alpha[1], rand(T,m), rand(T,m))
> > Inside the "testf" function, it copies the GPU arrays back to the CPU. Your question indicates that a synchronization is already there. Would you point us to the synchronization code?
> > The oneMKL library call always return a SYCL event, do we need to hook it up into oneAPI.jl's dependency tracking module?
> > Another scenario is that the spontaneous failure is caused by running out of memory that you have recently fixed, unlikely but possible.
> 
> I can see the queue synchronization command at: https://github.com/JuliaGPU/oneAPI.jl/blob/master/src/array.jl#L322
> 
> Let's run ze_tracer to see the sequence of L0 calls when it failed.

One suspect is the SYCL runtime batching, which waits and grouping zeCommandList in batches for submission to reduce L0 submission overhead:

https://github.com/intel/llvm/blob/sycl/sycl/doc/EnvironmentVariables.md

The environment variable SYCL_PI_LEVEL_ZERO_BATCH_SIZE=0 (default), letting the SYCL runtime to decide how many commands to wait before submitting them to the L0 queue.

When Julia issues zeSynchronizeCommandQueue, some of the oneMKL kernel could be waiting in the SYCL runtime to be batch-submitted.

**Comment by Sarbojit2019** [comment]
@kballeda, wait() is no longer needed as we have added force flush on every submission. Please update the PR with "__FORCE_MKL_FLUSH__(status)", you can find the use in latest source.

**Comment by kballeda** [comment]
Updated with force_mkl_flush instead of wait()

**Comment by maleadt** [comment]
LGTM. Needs a rebase though.

**Pull request merged**